fastestimator/op/tensorop/model/update.py

```diff
@@ -74,22 +107,130 @@ def fe_retain_graph(self, retain: Optional[bool] = None) -> Optional[bool]:
+            strategy = tf.distribute.get_strategy()
+            if isinstance(strategy, tf.distribute.MirroredStrategy):
+                loss = loss / strategy.num_replicas_in_sync
+
+        elif self.framework == "torch":
+            # scale up loss for mixed precision training to avoid underflow
+            if self.model.current_optimizer.scaler is not None:
+                loss = self.model.current_optimizer.scaler.scale(loss)
+
+        else:
+            raise ValueError(f"Unrecognized framework {self.framework}")
+
+        return loss
+
+    def _get_gradient(self, loss, state):
+        if self.framework == "tf":
+            gradients = get_gradient(loss, self.model.trainable_variables, tape=state['tape'])
+
+        elif self.framework == "torch":
+            trainable_params = [p for p in self.model.parameters() if p.requires_grad]
+            try:
+                gradients = get_gradient(loss, trainable_params, retain_graph=self.retain_graph)
+            except RuntimeError as err:
+                if err.args and isinstance(err.args[0], str) and err.args[0].startswith(
+                        'one of the variables needed for gradient computation has been modified by an inplace operation'
+                ):
+                    raise RuntimeError(
+                        "When computing gradients for '{}', some variables it relied on during the forward pass had"
+                        " been updated. Consider setting defer=True in earlier UpdateOps related to models which "
+                        "interact with this one.".format(self.model.model_name))
+                raise err
+
+        else:
+            raise ValueError(f"Unrecognized framework {self.framework}")
+
+        return gradients
+
+    def _gradient_postprocess(self, gradients, state):
+        """ Gradient postprocess for multi-GPU and mixed-precision training
+        """
+        if self.framework == "tf":
+            if self.gradients is not None:  # user provide gradients
+                strategy = tf.distribute.get_strategy()  # need tape.stop_recording() ?
+                # for multi-gpu training, the gradient will be combined by sum, normalize the gradient
+                if isinstance(strategy, tf.distribute.MirroredStrategy):
+                    gradients = gradients / strategy.num_replicas_in_sync
+
+            # scale down gradient to balance scale-up loss
+            if self.model.mixed_precision:
+                with state["tape"].stop_recording():
+                    gradients = self.model.current_optimizer.get_unscaled_gradients(gradients)
+
+        elif self.framework == "torch":
+            pass
+
+        else:
+            raise ValueError(f"Unrecognized framework {self.framework}")
+
+        return gradients
+
+    def _merge_grad_update(self, gradients, state):
+        current_grad = gradients
+        if self.framework == "tf":
+            self.step.assign_add(1)
```

Comment:
should we do modulo addition here rather than pure increase with mod later? Not sure whether speed of mod is dependent on numeric value of input

Reply:
Ya, just tested and confirmed that batch gradient is the sum of individual data gradient.

Reply:
I will make it use sum instead of average.

Reply:
I am not sure what you meant.
If you meant self.step increase to self.merge_grad value and goes back to zero and repeat, I already tried it and it has werid issues in tf case with GPU.

Reply:
interesting. How did you implement it when you ran into problem with tf gpu?

Reply:
for gs, g in zip(self.grad_sum, gradients):
          self._assign_add(gs, g)
      if self.step == self.merge_grad - 1:
          update_model(model=self.model, gradients=self.grad_sum, defer=self.defer, deferred=deferred)
          for gs in self.grad_sum:
              self._assign_add(gs, -gs)  # zero the gradient in place
          self._assign_add(self.step, self.step) # reset self.step to 0
      else:
          self._assign_add(self.step, 1)

Reply:
huh. do you still have the error message from that?

Reply:
Ya, I ran it last Friday.

Reply:
error message:
tensorflow.python.framework.errors_impl.InvalidArgumentError: Cannot place the graph because a reference or resource edge connects colocation groups with incompatible assigned devices: /job:localhost/replica:0/task:0/device:CPU:0 vs /job:localhost/replica:0/task:0/device:GPU:0. The edge src node is Func/cond/StatefulPartitionedCall/input/_24 , and the dst node is cond/StatefulPartitionedCall/cond [Op:__inference__forward_step_static_22597]